src/FSharp.Core/seq.fs

```diff
@@ -1938,3 +1938,186 @@ module Seq =
             if i < index then
                 invalidArg "index" "index must be within bounds of the array"
         }
+
+    [<CompiledName("RandomShuffleWith")>]
+    let randomShuffleWith (random: Random) (source: seq<'T>) : seq<'T> =
+        checkNonNull "random" random
+        checkNonNull "source" source
+
+        let tempArray = toArray source
+        Microsoft.FSharp.Primitives.Basics.Random.shuffleArrayInPlaceWith random tempArray
+        ofArray tempArray
+
+    [<CompiledName("RandomShuffleBy")>]
+    let randomShuffleBy (randomizer: unit -> float) (source: seq<'T>) : seq<'T> =
+        checkNonNull "source" source
+
+        let tempArray = toArray source
+        Microsoft.FSharp.Primitives.Basics.Random.shuffleArrayInPlaceBy randomizer tempArray
+        ofArray tempArray
+
+    [<CompiledName("RandomShuffle")>]
+    let randomShuffle (source: seq<'T>) : seq<'T> =
+        randomShuffleWith ThreadSafeRandom.Shared source
+
+    [<CompiledName("RandomChoiceWith")>]
+    let randomChoiceWith (random: Random) (source: seq<'T>) : 'T =
+        checkNonNull "random" random
+        checkNonNull "source" source
+
+        let inputLength = source |> length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        let i = random.Next(0, inputLength)
+        source |> item i
+
+    [<CompiledName("RandomChoiceBy")>]
+    let randomChoiceBy (randomizer: unit -> float) (source: seq<'T>) : 'T =
+        checkNonNull "source" source
+
+        let inputLength = source |> length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        let i = Microsoft.FSharp.Primitives.Basics.Random.next randomizer 0 inputLength
+        source |> item i
+
+    [<CompiledName("RandomChoice")>]
+    let randomChoice (source: seq<'T>) : 'T =
+        randomChoiceWith ThreadSafeRandom.Shared source
+
+    [<CompiledName("RandomChoicesWith")>]
+    let randomChoicesWith (random: Random) (count: int) (source: seq<'T>) : seq<'T> =
+        checkNonNull "random" random
+        checkNonNull "source" source
+
+        if count < 0 then
+            invalidArgInputMustBeNonNegative "count" count
+
+        let inputLength = source |> length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        seq {
+            for _ = 0 to count - 1 do
+                let j = random.Next(0, inputLength)
+                source |> item j
+        }
```

Comment:
I think `source` should be cached here, probably as an array, since otherwise it will be enumerated $count + 1$ times, which could result in very unpredictable behavior if enumerating it has side effects.

Reply:
I had this concern as well and it's a tough one... While in most cases it's true, there is one case when user might want to use iteration rather than copying array - when array is very big and count is pretty small and available memory is small. 

Also, when we cache all functions implementations to arrays, the question emerges - why do we even have random methods on sequence? Maybe we should just leave array and list versions and let user convert their sequence to array themselves?

I anticipated this problem in initial suggestion, but [Don wanted to add sequences as well](https://github.com/fsharp/fslang-suggestions/issues/508#event-5841982507)
![image](https://github.com/dotnet/fsharp/assets/3329606/82f9cf43-ed63-4302-81c7-70cacdb860a7)
so I decided to keep this ineffective version for consistency with other slow sequence operations (like `item`). However, I'm OK to remove all the sequence functionality at all.

Reply:
Ok, agree, iterating only once is probably what user will want in all cases. So added caching to array everywhere